plugins/wasmedge_stablediffusion/sd_func.cpp

```diff
@@ -272,47 +350,55 @@ Expect<uint32_t> SDTextToImage::body(
   std::string InputIdImagesDir(InputIdImagesDirSpan.begin(),
                                InputIdImagesDirSpan.end());
   std::string OutputPath(OutputPathSpan.begin(), OutputPathSpan.end());
-  if (!parameterCheck(Env, Width, Height, SessionId)) {
+  if (!parameterCheck(Env, Width, Height, SessionId, OutputPathLen)) {
     return static_cast<uint32_t>(ErrNo::InvalidArgument);
   }
   sd_ctx_t *SDCtx = Env.getContext(SessionId);
   sd_image_t *Results = nullptr;
   sd_image_t *ControlImage = nullptr;
-  uint8_t *ControlImageBuffer = nullptr;
   if (ControlImageLen != 0) {
     MEM_SPAN_CHECK(ControlImageSpan, MemInst, uint8_t, ControlImagePtr,
                    ControlImageLen,
                    "Failed when accessing the control image memory."sv)
-    ControlImage = readControlImage(ControlImageSpan, ControlImageBuffer, Width,
-                                    Height, CannyPreprocess);
+    ControlImage =
+        readControlImage(ControlImageSpan, Width, Height, CannyPreprocess);
   }
   spdlog::info("[WasmEdge-StableDiffusion] Start to generate image."sv);
   Results =
       txt2img(SDCtx, Prompt.data(), NegativePrompt.data(), ClipSkip, CfgScale,
               Guidance, Width, Height, sample_method_t(SampleMethod),
               SampleSteps, Seed, BatchCount, ControlImage, ControlStrength,
               StyleRatio, NormalizeInput, InputIdImagesDir.data());
-  // TODO upscale image
-  int Len;
-  unsigned char *Png = stbi_write_png_to_mem(
-      reinterpret_cast<const unsigned char *>(Results->data), 0, Results->width,
-      Results->height, Results->channel, &Len, nullptr);
-  if (OutputPathLen != 0) {
-    stbi_write_png(OutputPath.data(), Results->width, Results->height,
-                   Results->channel, Results->data, 0, nullptr);
-  }
-  *BytesWritten = Len;
+  free(ControlImage);
+  if (Results == nullptr) {
+    spdlog::error("[WasmEdge-StableDiffusion] Generate failed."sv);
+    free(SDCtx);
+    return static_cast<uint32_t>(ErrNo::RuntimeError);
+  }
+  // Upscale image
+  if (UpscaleModelPathLen != 0) {
+    MEM_SPAN_CHECK(UpscaleModelSpan, MemInst, char, UpscaleModelPathPtr,
+                   UpscaleModelPathLen,
+                   "Failed when accessing the Upscaler Image memory."sv)
+    std::string UpscaleModelPath(UpscaleModelSpan.begin(),
+                                 UpscaleModelSpan.end());
+    upscalerModel(UpscaleModelPath.data(), UpscaleRepeats, NThreads, Wtype,
+                  BatchCount, Results);
+  }
+  // Save results
+  unsigned char *Png = nullptr;
+  *BytesWritten = saveResults(Results, OutputPath, BatchCount, &Png);
   if (OutBufferMaxSize < *BytesWritten) {
     spdlog::error("[WasmEdge-StableDiffusion] Output buffer is not enough."sv);
+    free(SDCtx);
     free(Png);
     free(Results);
-    free(ControlImageBuffer);
     return static_cast<uint32_t>(ErrNo::RuntimeError);
   }
   std::copy_n(Png, *BytesWritten, OutputBufferSpan.data());
```

Comment:
copy_n is used to write png to OutputBuffer. Is moving it to the saveResults function more proper?